webapp/channels/src/components/admin_console/access_control/editors/cel_editor/editor.tsx

```diff
@@ -0,0 +1,363 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import * as monaco from 'monaco-editor';
+import React, {useCallback, useEffect, useRef, useState} from 'react';
+import {FormattedMessage} from 'react-intl';
+
+import type {AccessControlTestResult} from '@mattermost/types/access_control';
+
+import {searchUsersForExpression} from 'mattermost-redux/actions/access_control';
+import {Client4} from 'mattermost-redux/client';
+
+import {MonacoLanguageProvider} from './language_provider';
+
+import CELHelpModal from '../../modals/cel_help/cel_help_modal';
+import TestResultsModal from '../../modals/policy_test/test_modal';
+import {TestButton, HelpText} from '../shared';
+
+import './editor.scss';
+
+export const POLICY_LANGUAGE = 'expressionLanguage';
+const VALIDATE_POLICY_SYNTAX_COMMAND_ID = 'policyEditorValidateSyntaxCommand';
+
+const MONACO_EDITOR_OPTIONS: monaco.editor.IStandaloneEditorConstructionOptions = {
+    extraEditorClassName: 'policyEditor',
+    language: POLICY_LANGUAGE,
+    automaticLayout: true,
+    minimap: {enabled: false},
+    lineNumbers: 'off',
+    scrollBeyondLastLine: false,
+    wordWrap: 'on',
+    renderLineHighlight: 'none',
+    lineNumbersMinChars: 1,
+    occurrencesHighlight: 'off',
+    stickyScroll: {enabled: false},
+    autoClosingBrackets: 'never',
+    autoClosingQuotes: 'never',
+    autoIndent: 'keep',
+    autoSurround: 'never',
+    codeLens: false,
+    folding: false,
+    fontFamily: 'monospace',
+    hideCursorInOverviewRuler: true,
+    fontSize: 12,
+    guides: {indentation: false},
+    links: true,
+    matchBrackets: 'never',
+    multiCursorLimit: 1,
+    overviewRulerBorder: false,
+    quickSuggestions: false,
+    renderControlCharacters: false,
+    scrollbar: {
+        horizontal: 'hidden',
+        useShadows: false,
+    },
+    selectionHighlight: false,
+    showFoldingControls: 'never',
+    suggestOnTriggerCharacters: true,
+    unicodeHighlight: {
+        ambiguousCharacters: false,
+        invisibleCharacters: false,
+    },
+    unusualLineTerminators: 'auto',
+    wordWrapColumn: 400,
+    wrappingIndent: 'none',
+    wrappingStrategy: 'advanced',
+    contextmenu: false,
+};
+
+interface CELEditorProps {
+    value: string;
+    onChange: (value: string) => void;
+    onValidate?: (isValid: boolean) => void;
+    placeholder?: string;
+    className?: string;
+    userAttributes: Array<{
+        attribute: string;
+        values: string[];
+    }>;
+}
+
+// TODO: this is just a sample schema for the editor, we need to get the actual schema from the server
+
+function CELEditor({
+    value,
+    onChange,
+    onValidate,
+    placeholder = 'user.attributes.<attribute> == <value>',
+    className = '',
+    userAttributes,
+}: CELEditorProps): JSX.Element {
+    const [editorState, setEditorState] = useState({
+        expression: value,
+        isValidating: false,
+        isValid: true,
```

Comment:
(Post-merge) Do we need all 3 of `isValidating`, `isValid`, and `validationErrors`? It seems like `isValid` could be equal to `!isValidating && validationErrors.length == 0`, and that'd help avoid this somehow getting into a weird state by mistake where it's somehow valid and validating and invalid the same time